### PR TITLE
fv3fit: add layer clipping and implement as `dense` model option

### PR DESCRIPTION
**Pull request by AnnaKwa**
This PR adds a `ClipConfig` class to be used in keras models for clipping along the feature dimension (assumed to be the last dim) and implements this as an option in the `dense` model. Inputs are clipped by adding a clipping layer in the model. The output layers of the model used to fit data (`train_model`) are clipped, as well as the output data passed to the `fit` method. The output layers of the model used for prediction (`predict_model`) are of the same feature dimension size as the unclipped data, but with the clipped levels filled with zeroes.

I chose to implement all the clipping and refilling with zeros as layers within the keras model so that using a `Predictor` with clipped inputs/outputs would not require any knowledge of the information in the clip config. This comes at the expense of making the training and `build_model` functions more complex, with various helper functions included to clip both layers (when building model) and output data arrays (when passing training data to the model during fitting).

Added public API:
- `clip_config` optional field in the `DenseHyperparameters` to specify how to clip inputs and outputs. Loading the old `DenseModel` models that called this `packer_config` is backwards compatible.

- `XyMultiArraySequence` can take an optional arg `output_clip_config` for slicing the output data used in training (to match the dimensions of clipped model outputs).

- [x] Tests added

### CODE REVIEW

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+                array.take(
+                    indices=range(start, stop), axis=-1  # type: ignore
+                )
+            )
+        else:
+            outputs.append(array)
+    return outputs
+
+
+def _zero_slice_with_placeholder_dim(layer: tf.Tensor, n_zero_levels: int) -> tf.Tensor:
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/keras/_models/shared/clip.py:151]: If there is a more concise way to create a zeroes layer with a placeholder first dimension I would love to hear it. Stackoverflow seems to suggest using `tf.placeholder` but that is not available in the version of Tensorflow that we use.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 2]
```diff
@@ -84,7 +113,8 @@ def train_dense_model(
             dataset_sequence=[validation_batches[0]],
             unstacked_dims=["z"],
             n_halo=0,
-        )[0]
+        )
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/dense.py:116]: Does the code not break if you pass a sequence of (X, y) pairs as validation_data to `model.fit`? If it doesn't, this is definitely better, but I'm surprised.

Keep in mind that the model evaluates on the validation data every batch, so this could slow down training considerably depending on the case. Mostly something to keep in mind when you're configuring runs. Also you may want to communicate this to the team as a backwards-incompatible change (before it ignored all but the first batch).

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 3]
```diff
@@ -95,9 +125,16 @@ def train_dense_model(
...
         n_halo=0,
     )
+    # arrays used in build_model should be the full inputs and outputs before clipping
+    X, y = train_data[0]
+
+    # y train data should be clipped when used as target to fit train_model
+    train_data, validation_data = _clip_data_outputs(
+        hyperparameters, train_data, validation_data  # type: ignore
+    )
+
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/dense.py:135]: Using `validation_data[0]` before meant validation data is always in-memory, which makes sense because it's used for every batch. Can you add a cast of validation_data to `tuple` somewhere, which occurs no matter what (unlike train_data which is cast conditionally)?

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        # 1D layer that are REMOVED by clipping. Used to pad output layer back to their
+        # original size.
+        variable_slice_config = list(self.clip[name].values())[0]
+        start = variable_slice_config.start or 0
+        stop = variable_slice_config.stop or original_1d_layer_size
+        n_removed_from_start = start
+        n_removed_from_end = original_1d_layer_size - stop
+        return (n_removed_from_start, n_removed_from_end)
+
+    def clip_layer_along_feature_dim(self, layer, name):
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/keras/_models/shared/clip.py:31]: These are all public APIs - can you please add type hints and docstrings where they're missing?

nit: can you rename this to `clip_layer_along_last_dim`? Technically speaking, all non-sample dimensions are feature dimensions. We've been using "feature dimension" to mean the last dimension even if some feature dimensions are x/y instead of z, which is OK conversationally but we should be careful not to put too much in the code when there's a clearer alternative.

File: external/fv3fit/fv3fit/keras/_models/dense.py (hunk 2 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:116]: Interestingly, it does not.

Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:116]: Actually there was another line in addition to this that ensured only the first item in the sequence was used (ln 113 above).

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                    indices=range(start, stop), axis=-1  # type: ignore
+                )
+            )
+        else:
+            outputs.append(array)
+    return outputs
+
+
+class ClippedXyMultiArraySequence(tf.keras.utils.Sequence):
+    def __init__(
```
Comment (frodre) [inline external/fv3fit/fv3fit/keras/_models/shared/clip.py:169]: Mentioned above that we might just consider adding clipping as an option to `XyMultiArraySequence`.  Although, it does shorten the code for grabbing the "full" fields and clipped fields.

**Review by frodre — CHANGES_REQUESTED** [review]
The great clippening! Thanks @AnnaKwa for the PR, I have a few comments I left in the code that could help streamline.  Comment below is born out of me also attempting this w/ emulation previously and ending up in a similar spot but not being totally happy.

A potential major change to consider is whether more can be packed into the tensorflow logic.  Right now, most of the in and outside model interactions are based around the need to have the training properly consider the right layers in the loss but output full fields (w/ padded zeros).  The interactions are currently facilitated through quite a few function calls adjusting the training data and having two separate models.  

I was thinking that we could maybe achieve the same end w/o having to bookkeep the training data by either taking advantage of weights or custom loss functions.  

1. Weights could be provided in the custom training function to just ignore the clipped layers. Since the model enforces zero at those levels we wouldn't have to worry about "untrained" predictions.
2. Could also just use custom loss functions with the clipping info embedded to enforce the constraints 

But ... if either of these ideas don't actually simplify the implementation and understanding, then it's probably not worth it.  Or if this has already been discussed and disqualified for reasons, A-OK.  My main concerns around this are with maintainability of the Dense model with the added training complexity.  Happy to chat about anything mentioned though if it's quicker than the PR back and forth.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 6]
```diff
@@ -62,15 +67,39 @@ class DenseHyperparameters(Hyperparameters):
...
+    return ClipConfig(output_only_config)
+
+
+def _clip_data_outputs(
+    hyperparameters: DenseHyperparameters,
+    train_data: XyMultiArraySequence,
+    validation_data: Optional[XyMultiArraySequence] = None,
+):
+    # fit loop needs to clip the y target feature dim if train_model outputs are clipped
+    y_only_clip_config = _clip_config_outputs_only(
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:91]: In the current way the clipping is implemented, the `trained_model` takes the full (unclipped) inputs and outputs the clipped outputs, so only the y's are clipped in the array data. I like your suggestion better though to clip outputs in training via the loss weights, so this part will be removed.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 7]
```diff
@@ -62,15 +67,39 @@ class DenseHyperparameters(Hyperparameters):
...
+        hyperparameters.clip_config, hyperparameters.output_variables
+    )
+    y_clipped_train_data = ClippedXyMultiArraySequence(train_data, y_only_clip_config)
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:94]: I like the suggestion to use the loss weights for output clipping instead, which will remove any need to clip the training data, so I'm removing this.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 8]
```diff
@@ -167,16 +209,38 @@ def build_model(
             tf.keras.layers.Activation(tf.keras.activations.relu)(output_layer)
             for output_layer in denorm_output_layers
         ]
-    train_model = tf.keras.Model(inputs=input_layers, outputs=denorm_output_layers)
+    clipped_denorm_output_layers = clip_layers(
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py:212]: I'm changing output clipping to use the weights instead, which should simplify this a lot more and remove these lines.

That is a good point about the uniform scaling option in the loss being dependent on the output clipping range- I'll make sure it is using the clipped range only in the final implementation.

File: external/fv3fit/fv3fit/keras/_models/dense.py [hunk 9]
```diff
@@ -167,16 +209,38 @@ def build_model(
...
+        config.clip_config, list(output_stds), config.output_variables
+    )
     train_model.compile(
         optimizer=config.optimizer_config.instance,
-        loss=[config.loss.loss(std) for std in output_stds],
+        loss=[config.loss.loss(std) for std in clipped_output_stds],
+    )
+
+    # need a separate model for this so i) we don't have to
+    # serialize the custom loss functions and ii) clipped outputs can be
```
[the remark this answers is not in the document]
Comment (nbren12) [reply external/fv3fit/fv3fit/keras/_models/dense.py:232]: Yah. these tensorflow apis are really poorly documented. Took me about 30min-hour of debugging to figure that out.

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+                variable_slice_config.start,
+                variable_slice_config.stop,
+            )
+            feature_start = feature_start or 0
+            feature_stop = feature_stop or layer.shape[-1]
+
+            slice_start = [0 for l in layer.shape[:-1]] + [feature_start]
+            slice_size = [-1 for l in layer.shape[:-1]] + [feature_stop - feature_start]
+
+            return tf.slice(layer, slice_start, slice_size)
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/clip.py:50]: Thanks, I didn't know you could refer to a slice in a layer like that. That way is much cleaner.

File: external/fv3fit/fv3fit/keras/_models/dense.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/dense.py]: Update: I know we discussed implementing this, but I ended up reverting to the original approach of slicing to the final range of outputs used in training. The reason for this is that the treatment of the standard deviation in the loss was getting too gnarly and it felt like pushing these details into a lower level of the training code would make later alterations trickier.

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        if name in config.clip:
+            outputs.append(config.clip_layer_along_feature_dim(layer, name))
+        else:
+            outputs.append(layer)
+    return outputs
+
+
+def clip_arrays(
+    config: ClipConfig, arrays: Sequence[np.ndarray], variable_names: Sequence[str],
+) -> Sequence[np.ndarray]:
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/clip.py]: Thanks for the comment on slicing, these are now merged to `clip_sequence`.

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+class ClipConfig(PackerConfig):
+    clip: ClipDims = dataclasses.field(default_factory=dict)
+
+    def single_feature_slice(self, name: str) -> SliceConfig:
+        self._check_only_feature_dim_is_clipped(name)
+        variable_slice_config = list(self.clip[name].values())[0]
+        return variable_slice_config
+
+    def removed_sizes(self, original_1d_layer_size: int, name: str):
+        # Returns the size of the feature dim slices at the start and end of a
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/clip.py]: I think that so far we have only ever used this on the last (vertical) dimension of the training data although the original additional left the possibility open for clipping other dimensions as well.

**Comment by AnnaKwa** [comment]
Thanks for the feedback @frodre, it made the code a lot more concise, this is ready for re-review.

I know we discussed the output clipping implementation earlier before the holiday break and landed on using 1/0 weight masking in a single model for both train and prediction. After working on those changes I ended up return to a similar implementation as originally written, with separate train/predict models with clipped/nonclipped output layer sizes. My reasoning for this was 

1) I'd have to go pretty deep to correctly calculate the treatment of the 'uniform standard' weighting in the loss if there were zeroed output levels, while dealing with this is trivial if you exclude those levels when calculating the standard dev. Although there multiple places in the training function and model building that have to call clip functions, I think it is easier to follow what's going on if the clipping logic is higher up in the model training/building code.

2) I think the pattern of having different train/predict models could still be used in the future if we use additional outputs in the loss that aren't directly used in prediction

I did end up using the zero weighting to produce the predict model output layers as opposed to concatenating zeros onto the clipped outputs; this was much cleaner to implement and avoids the messy helper functions in `clip.py`.

**Review by frodre — APPROVED** [review]
Thanks for giving it a shot to bringing everything into the model @AnnaKwa.  Good to know where the pain points still are for doing that.  I think the updates have really cleaned things up and it looks like it'll be a lot easier to maintain in that regard! Nice work.

> I think the pattern of having different train/predict models could still be used in the future if we use additional outputs in the loss that aren't directly used in prediction

Yep that's a valid use.  We have extra variables in the emulator right now, but since they don't overlap with anything important name-wise it's okay to have that single model.

Only question is the usage of a config that allows for clip dimension specification but then doesn't really use it.  Is there a planned use case for it, or is it a vestige of some other process?

File: external/fv3fit/fv3fit/keras/_models/shared/sequences.py [hunk 13]
```diff
@@ -76,6 +79,11 @@ def __init__(
                 arrays apart from the "sample" (stacked) dimension
             n_halo: number of halo points to append to input variables, if given
                 then the data must contain "x" and "y" dimensions
+            output_clip_config: If provided, will zero mask out levels of output
+                variables that are clipped in the clip config when returning arrays.
+                Note that this will not affect input variables, even if they are in
+                the clip config, as we want to input them in their full length when
+                training (even if they are clipped in a subsequent layer).
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/sequences.py:86]: Correct, thanks for catching that, it was left over from the previous iteration of changes that got reverted.

File: external/fv3fit/tests/test_clip.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+            np.array([0, 1]),
+            id="clip_1d_no_start",
+        ),
+        pytest.param(
+            ClipConfig({"var_": {"z": SliceConfig(8, None)}}),
+            np.arange(10),
+            id="clip_var_not_in_config",
+        ),
+        pytest.param(
+            ClipConfig({"var": {"z": SliceConfig(8, None), "y": SliceConfig(None, 2)}}),
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/tests/test_clip.py:33]: I went with the suggestion in another comment to just make the class assume that you are only going to clip one dimension, and fail to initialize if not, so I removed this test case.

File: external/fv3fit/fv3fit/keras/_models/shared/clip.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        """Fills clipped levels with zero, maintaining the original length along the
+        clipped dimension. If name is not in config, returns the input layer unchanged.
+
+        Args:
+            layer: tensor/layer to clip along last dim
+            name: variable name corresponding to entry in ClipConfig.clip
+        """
+        if name in self.clip:
+            mask = self._get_mask_array(layer, name)
+            mask_layer = tf.transpose(tf.constant(mask, dtype=tf.float32))
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/keras/_models/shared/clip.py:37]: It wasn't doing anything, so I removed it.
